python/mspasspy/db/database.py

```diff
@@ -2761,6 +2769,336 @@ def read_ensemble_data(
+                    # had an ancestor that had an option to read data here.
+                    # we never do that here
+                    mspass_object = Seismogram(md, False)
+                    ensemble.member.append(mspass_object)
+            except MsPASSError as merr:
+                # if the constructor fails mspass_object will be invalid
+                # To preserve the error we have to create a shell to hold the error
+                if object_type is TimeSeries:
+                    mspass_object = TimeSeries()
+                else:
+                    mspass_object = Seismogram()
+                # Default constructors leaves result marked dead so below should work
+                mspass_object.elog.log_error(merr)
+                return mspass_object
+
+                # not continue step 2 & 3 if the mspass object is dead
+            if is_dead:
+                mspass_object.kill()
+                for msg in log_error_msg:
+                    mspass_object.elog.log_error(
+                        "read_data", msg, ErrorSeverity.Invalid
+                    )
+            else:
+                # 2.load data from different modes
+                mspass_object.live = True
+                # 3.load history
+                if load_history:
+                    history_obj_id_name = (
+                        self.database_schema.default_name("history_object") + "_id"
+                    )
+                    if history_obj_id_name in object_doc:
+                        self._load_history(
+                            mspass_object,
+                            object_doc[history_obj_id_name],
+                            alg_name,
+                            alg_id,
+                            define_as_raw,
+                            retrieve_history_record,
+                        )
+
+                mspass_object.clear_modified()
+
+                # 4.post complaint elog entries if any
+                for msg in log_error_msg:
+                    mspass_object.elog.log_error(
+                        "read_data", msg, ErrorSeverity.Complaint
+                    )
+
+        res = cur_collection.aggregate(
+            [
+                {"$match": {"$expr": {"$in": ["$_id", objectid_list]}}},
+                {
+                    "$group": {
+                        "_id": {"dir": "$dir", "dfile": "$dfile"},
+                        "foffs": {"$push": "$foff"},
+                        "ids": {"$push": "$_id"},
+                    }
+                },
+            ]
+        )
+
+        files = list(res)
+        for f in files:
+            cur_dir = f["_id"]["dir"]
+            cur_dfile = f["_id"]["dfile"]
+            foofs = list(map(int, f["foffs"]))  # string list -> int list
+            # get original index
+            indexes = list(map(objectid_list.index, f["ids"]))
+
+            try:
+                cnt = _fread_from_files(
```

Comment:
This couldn't have worked before.   You need to first build the skeleton of the ensemble before call the new C++ function `_fread_from_file`.   Suggest this be done as a helper, private method.   The reason is you can probably write a single function to handle building the skeleton for both the Seismogram and TimeSeries versions.  Use the constructors for both objects that take a `Metadata` and `BasicTimeSeries` as inputs.  You create the Metadata container directly from mongodb document retrieved in a cursor loop, construct the BasicTimeSeries object, and then call that constructor.  That can be done in the cursor loop pushing each skeleton to the member container.     If that isn't clear (highly likely) reply and I'll try to expand.

Reply:
I don't quite understand here. I think I have built the ensemble skeleton in the above code (from line 2855), which is copied from the `read_data` function. Firstly build the `Metadata` and then use metadata to build the `TimeSeries` or `Seismogram` object, next add the object to the ensemble. And when this line `_fread_from_files` is executed, there are already objects in the ensemble. I will add more comments in this function to explain the confusing part.